### ISSUE TITLE
Cannot connect to gateway on a host with many network interfaces by setting EPICS_PVA_ADDR_LIST

### ISSUE THREAD

**Issue by goetzpf** [epics-base/p4p#162]
I have the following setup:

- The Gateway host has several interfaces.
- The pva gateway is configured to run the server on the second network interface
- The client is in a separate network that has only a route to this second interface
- The client connects by setting EPICS_PVA_ADDR_LIST

When I use EPICS_PVA_ADDR_LIST to connect directly to the gateway, the repsonse to the channel discovery request contains the wrong source interface address, not the one that was selected with EPICS_PVA_ADDR_LIST.

Here is the concrete example:

The gateway host has (among others) these interfaces;

> $ ip -4 addr
> 1: lo: <LOOPBACK,UP,LOWER_UP> mtu 65536 qdisc noqueue state UNKNOWN group default qlen 1000
>     inet 127.0.0.1/8 scope host lo
>        valid_lft forever preferred_lft forever
> 10: bond0: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
>     inet 172.30.134.57/24 brd 172.30.134.255 scope global bond0
>        valid_lft forever preferred_lft forever
> 11: bond3: <BROADCAST,MULTICAST,MASTER,UP,LOWER_UP> mtu 1500 qdisc noqueue state UP group default qlen 1000
>     inet 192.168.16.7/24 brd 192.168.16.255 scope global bond3
>        valid_lft forever preferred_lft forever

This is the gateway configuration file:

> {
>     "version":2,
>     "clients":[
>         {
>             "name":"GWCLIENT1",
>             "provider":"pva",
>             "addrlist":"172.30.130.255",
>             "autoaddrlist":false,
>             "bcastport":5076
>         }
>     ],
>     "servers":[
>         {
>             "name":"CASERVER1",
>             "clients":["GWCLIENT1"],
>             "interface":["192.168.16.7"],
>             "addrlist":"192.168.16.255",
>             "autoaddrlist":false,
>             "serverport":5075,
>             "bcastport":5076,
>             "getholdoff":0,
>             "statusprefix":"PV:",
>             "access":"",
>             "pvlist":""
>         }
>     ]
> }

I run this command on the client host (ip 193.149.12.42):

> $ (export EPICS_PVA_DEBUG=100; export EPICS_PVA_AUTO_ADDR_LIST=NO; export EPICS_PVA_ADDR_LIST=192.168.16.7; pvinfo idadm:calc1)
> 2024-12-05T16:29:14.637 Creating datagram socket from: 0.0.0.0:58085.
> 2024-12-05T16:29:14.637 Broadcast address #0: 192.168.16.7:5076. (unicast)
> 2024-12-05T16:29:14.637 Setting up UDP for interface 193.149.12.42/255.255.255.0, broadcast 193.149.12.255, dest <none>.
> 2024-12-05T16:29:14.637 Creating datagram socket from: 193.149.12.42:5076.
> 2024-12-05T16:29:14.637 Creating datagram socket from: 193.149.12.255:5076.
> 2024-12-05T16:29:14.637 Creating datagram socket from: 224.0.0.128:5076.
> 2024-12-05T16:29:14.637 Local multicast enabled on 127.0.0.1/224.0.0.128:5076.
> 2024-12-05T16:29:14.637 Sending 57 bytes 0.0.0.0:58085 -> 192.168.16.7:5076.
> 2024-12-05T16:29:14.638 UDP Client Rx (53) 0.0.0.0:58085 <- 172.30.134.57:5076
> ...

The relevant information is this:

EPICS_PVA_ADDR_LIST is set to connect directly to 192.168.16.7. The response UDP package has a source address of 172.30.134.57, not 192.168.16.7 as it should be. In our network setup, the client host has a route to 192.168.16.7  but not to 172.30.134.57 so it cannot connect.

I have verified that the same command line works when I connect to a soft-IOC on a host with many interfaces with was started by setting EPICS_PVAS_INTF_ADDR_LIST only on one of the network interfaces.

So this must be a problem of the pva gateway and not the pv access implementation in EPICS base.

**Comment by mdavidsaver** [comment]
@goetzpf As you describe it, your configuration seems correct.  Which makes me think that something more is going on.

- Can you provide your routing table?  `ip -4 route`
- Can you access any gateway status PV?  eg [`pvget PV:clients`](https://epics-base.github.io/p4p/gw.html#status-pvs)
- In addition to `pvinfo` (from pvAccessCPP), please also test with `pvxinfo` (from PVXS).  This shouldn't make a difference, but is a useful data point.
- With `pvxinfo`, instead of setting `EPICS_PVA_ADDR_LIST`, instead set `EPICS_PVA_NAME_SERVERS=192.168.16.7` to avoid UDP entirely.

If none of these give any hints, could you repeat your test while running a packet capture on the gateway host?

eg. `tshark -i lo -i bond0 -i bond3 -w capture.pcapng`  You could then transfer `capture.pcapng` to another host and [filter only the `pva` traffic](https://epics-base.github.io/pvxs/details.html#packet-capture).

**Comment by mdavidsaver** [comment]
Something I should have asked earlier.

What all PVA peers (clients and servers) are running on the gateway host?  Which module version(s) are involved?

If you are using P4P older than 4.0.0, please re-test with a newer version.

> I have added a capture "raw-blc.pcapng.gz" ...

hmmm...  The unicast search packets being re-send as local multicast (packets 62481 and 63761) do not look correct in a way which I don't think that PVXS is capable of.  The "ORIGIN" address is `192.168.16.255` instead of the actual `192.168.16.42`.  Which tells me that some PVA peer using (probably) pvAccessCPP is receiving the unicast.

fyi. The local mulicast "hack" which lets PVA avoid the need for separate UDP ports like CA depends on all peers on a system cooperating.

**Comment by goetzpf** [comment]
> Something I should have asked earlier.
> 
> What all PVA peers (clients and servers) are running on the gateway host? Which module version(s) are involved?

The gateway currently only runs my PVA gateway. However, it has many CA-Gateway processes.
> 
> If you are using P4P older than 4.0.0, please re-test with a newer version.

I use the newest version from PyPI, 4.2.0.
> 
> > I have added a capture "raw-blc.pcapng.gz" ...
> 
> hmmm... The unicast search packets being re-send as local multicast (packets 62481 and 63761) do not look correct in a way which I don't think that PVXS is capable of. The "ORIGIN" address is `192.168.16.255` instead of the actual `192.168.16.42`. Which tells me that some PVA peer using (probably) pvAccessCPP is receiving the unicast.

Where do you see `192.168.16.255` and `ORIGIN` in the data ? 
Since we intend to run several PVA gateways on a single host, we have IPTABLES rules that convert UDP unicasts to 192.168.16.7 to local UDP broadcasts:
```
$ iptables -t nat -L -n
DNAT       17   --  0.0.0.0/0            172.30.134.57        udp dpt:5076 to:172.30.134.255
DNAT       17   --  0.0.0.0/0            192.168.16.7         udp dpt:5076 to:192.168.16.255
...
```
This ensures (at least so it works with the CA gateway) that all processes listening to the same IP address (here 192.168.16.7) and the same port (here 5076) get the UDP packet.
In my test there is currently  only one PVA gateway running.

> 
> fyi. The local mulicast "hack" which lets PVA avoid the need for separate UDP ports like CA. It does depends on all peers on a system cooperating.

What is this multicast hack ? Could it be that this is responsible for the wrong source IP ?

**Comment by goetzpf** [comment]
> > ... we have IPTABLES rules that convert UDP unicasts to 192.168.16.7 to local UDP broadcasts: ...
> > ... This ensures (at least so it works with the CA gateway) ...
> 
> This should **not** be done for PVA.

Okay, I have now removed the rules on our gateway host, verified with this command:
```
# iptables -t nat -L  | grep 5076
# iptables  -L  | grep 5076
```
> 
> > What is this multicast hack ? Could it be that this is responsible for the wrong source IP ?
> 
> The short answer is that this is the reason why you need not (and should not) do any mangling with firewall rules.
> 
> The slightly longer answer is that when a PVA peer (client or server) receives a UDP unicast message, it prepends a `CMD_ORIGIN_TAG` message, then re-sends as a multicast on the loopback interface. This multicast is received by all PVA peers, including the sender, which then process that it.
> 
> "hack" is my label as I don't like this aspect of the PVA design, which seems to me to be brittle.

Ich have repeated my test. The output of tshark is here:
https://e.pcloud.link/publink/show?code=XZEq3wZN22REXwj31FHtNAg5GkOQpeDLj9X

The screenshot below shows, that the source address is still wrong:
![Screenshot_2025-01-09_15-27-29](https://github.com/user-attachments/assets/37ab8229-867e-4754-b39d-1a425e92e807)

**Comment by goetzpf** [comment]
I tried to look more into the code what happens when the SEARCH_REPONSE is created. 
The problem is that in pvxs/src/udp_collector.cpp in the constructor of UDPCollector:
```
UDPCollector::UDPCollector(UDPManager::Pvt *manager, int af, uint16_t requested_port)
    :manager(manager)
    ,bind_addr(SockAddr::any(af, requested_port))      // <--- problem
    ,lo_mcast_addr("224.0.0.128,1@127.0.0.1")
    ,lo_addr(SockAddr::loopback(bind_addr.family()))
    ,sock(af, SOCK_DGRAM, 0)
    ,rx(__FILE__, __LINE__,
        event_new(manager->loop.base, sock.sock, EV_READ|EV_PERSIST, &handle_static, this))
    ,beaconMsg(src)
```
that the socket is bound to SockAddr::any which is in IPV4  0.0.0.0.

According to this source:
  https://blog.cloudflare.com/everything-you-ever-wanted-to-know-about-udp-sockets-but-were-afraid-to-ask-part-1/#sourcing-packets-from-a-wildcard-socket
and some others I found this means that the kernel tries to guess the source IP for the UDP packet that is sent.

This is wrong, the source IP *must* be the IP of the network interface where the PV Access Search Request was received.

I have looked if I can fix this myself but currently I'm stuck with finding the source IP of the network interface where the request was received. 

Maybe you, as the author, could fix this faster ?

**Comment by mdavidsaver** [comment]
> This is wrong, the source IP _must_ be the IP of the network interface where the PV Access Search Request was received.

I'm not sure that I agree with "must" in this assertion, although I take your point that aside from code complexity there should be little practical downside to short circuiting the routing table when replying to SEARCH.  I agree that it would help in a situation like this where the routing table is incomplete, but the ARP cache (probably) still contains an entry for the reverse path.

Switching from `sendto()` to `sendmsg()` and `IP_PKTINFO` would I think be the approach since `recvmsg()` with `IP_PKTINFO` are already is use when receiving SEARCH requests.  This will necessitate some winsock compatibility testing, although I suspect this will work.

> kernel tries to guess the source IP for the UDP packet that is sent.

There is no guessing.  The routing table fully determines this behavior.  Regardless of changes to PVXS, I must (regrettably) recommend making your already very large gateway host routing table even larger with an entry for `193.149.12.0/24`. eg. assuming that 192.168.16.1 is the appropriate router.

```sh
ip route add 193.149.12.0/24 via 192.168.16.1
```

Then run:

```sh
ip route get 193.149.12.42
```

If successful this should print an entry from the routing table like: `193.149.12.42 via 192.168.16.1 dev bond3 ...`.

**Comment by mdavidsaver** [comment]
> This will necessitate some winsock compatibility testing, although I suspect this will work.

I spoke too soon, this will take some thought.  All targets can provide IP "Header destination address" and receiving "Interface index" via `recvmsg()`.  Only Linux gives the additional `in_pktinfo::ipi_spec_dst` "Routing destination address", which is I think exactly what would be needed for "reply to".  A lookup from interface index to addresses is possible, but this may be ambiguous since each interface may have many addresses (for IPv6 this is the norm).

And there is also a complication that the "multicast hack" described above means that replies to a unicast are done based on a forwarded packet, where the `in_pktinfo` is not relevant for a reply to the original unicast.  This leaves the address from `CMD_ORIGIN_TAG`, which should be the IP "Header destination address" address.  This is what we want for a unicast, although not all PVA implementations inject CMD_ORIGIN_TAG.

That's tying my brain in enough of a knot tonight...

**Comment by goetzpf** [comment]
It seems that the source-ip of the SEARCH-RESPONSE UDP packet comes indeed from the routing set up for the receiver of the packet, the host that issued the request. 
However, if you have several gateway processes running on a host with many interfaces, this cannot be correct in general. For example, requests from networks where the gateway host has no interface are always replied to via the default IP gateway. This results in the same source IP in every case even if different PVA gateways with different server IP addresses  are replying. 

I think the UDP reply is generated in file pvxs/src/server.cpp in this function:
```
  void Server::Pvt::onSearch(const UDPManager::Search& msg)
```
Would it be possible to store the receiving network interface in the message object (UDPManager::Search) ?

I found a discussion about a similar problem here:
https://stackoverflow.com/questions/5755265/how-to-set-source-address-when-sending-using-and-udp-socket

This:
http://linux-ip.net/html/routing-saddr-selection.html
suggests one could use bind() to set the source IP address.

**Comment by mdavidsaver** [comment]
> ... gateway host has no interface are always replied to via the default IP gateway.

This is not a correct statement.  Your gateway host can and should have a complete routing table, with entries matching all reachable subnets.  I hope you will not consider consider me to be patronizing if I suggest that you do some reading about the IP routing process to understand what "matching" means in this context.  There is also an interplay with the ARP process which is less important to understand in that it should Just Work if all nodes have a complete routing table.

To stay in scope for this discussion, I believe that adding a limited `via` route (aka. a non-default IP gateway) will resolve your issue.  (see https://github.com/epics-base/p4p/issues/162#issuecomment-2638315479)

So your next task is to identify which node on `192.168.16.0/24 ` is acting as a router.  I am certain that there is one since your client host is able to route through `193.149.12.1` to reach `192.168.16.7`.  I _guess_ that this will be `192.168.16.1`.  Your network administrator will know.

**Comment by goetzpf** [comment]
Our gateway does not have an interface in 193.149.12.0/24. So all traffic to 193.149.12.42 must go through the default IP gateway, currently 172.30.134.1. This causes the source IP to be set to 172.30.134.57.
So it seems that the IP gateway used for routing to *any* client of a PVAccess gateway must be in the same network as the server interface of the PVAccess gateway.
Maybe I can set up routing to work in *this* case. 
However, if I have two PVA Gateways with server interfaces in different networks, both only connected to a client by a single default IP gateway, this will not work, connections will only be possible to *one* of them. I am not sure if this will be a problem for our setup.

**Comment by mdavidsaver** [comment]
I feel that I must be missing some critical detail about this situation.

> ... all traffic to 193.149.12.42 must go through the default IP gateway, currently 172.30.134.1. This causes the source IP to be set to 172.30.134.57.  ...

How do packets from `193.149.12.42` reach the `192.168.16.7` interface?

The provided packet capture file shows original search packet arriving on `bond3`.  If it comes through `172.30.134.1` it should arrive at the gateway host on `bond0`, right?

Is this PVA gateway host itself acting as an IP router?

Can you provide a complete iptables/nft listing?  (by private email if you prefer)

**Comment by goetzpf** [comment]
172.30.134.1 is the default IP gateway on the PVA gateway host. This host has no interface in network 193.149.12.0/24. The gateway host itself is not a router.

The client sends a packet to the PVA gateway by setting environment variables, this is the command I run:
```
EPICS_PVA_AUTO_ADDR_LIST=NO EPICS_PVA_ADDR_LIST=192.168.16.7 pvinfo idadm:calc1
```
I could create an iptables listing, however, I removed all extra rules for PVAccess ports which I had in the meantime, so this shouldn't be the problem now.

Setting EPICS_PVA_ADDR_LIST is not our usual mode how we operate the gateways but this should work, too.

Our usual setup with the CA-Gateways is in the following picture. We have had this in production for many years now with about 100 networks and only two gateway hosts, each with an interface in about one half of the networks.

![Image](https://github.com/user-attachments/assets/8073a844-79e9-4238-ad0a-35d142156291)

A small program. "udp broadcaster" forwards UDP broadcasts from all subnets to the CA network, where all CA gateway server interfaces are. 
A client name resolution request is forwarded as a UDP broadcast to the CA network. We have a two machines there with many network interfaces and many CA gateway processes. The CA gateway that has the PV sends a UDP reply packet to the client who, by examining the source IP address of the packet, then knows to which IP address it has to connect. The client then makes a direct connection to the CA gateway in the CA network.
By this setup, we have *one exporting* CA gateway for each PV that serves all clients in all other networks. We have the access security configuration for each PV in only *one file*. We do not depend on setting environment variables like EPICS_CA_ADDR_LIST on the client side.

I thought I would be able to set up PVA Gateways in a similar way, but since the UDP source address of the PVA gateway reply package depends on the routing configuration on the gateway host, this whole concept will no longer work.

In my example with which I started this issue, the PVA gateway has no interface in the network of the client, so I think it would work by changing the default route on the gateway host to the network where the PVA Gateway server address is, 192.168.16.1 in this case. But I think this is not a solution in the general case.

Consider a client that is in a network, where the PVA gateway has an interface, too, slightly different IP addresses here, but I saw the same problem on this second gateway host. So there is a routing entry on the gateway host for subnet1 that uses the IP address the gateway host has in subnet1, for example 
```
193.149.12.0/24 dev bond1.400 proto kernel scope link src 193.149.12.6
```
This means that the source IP of the UDP reply is not the IP address of the PVA gateway `192.168.16.6` in the CA network, but that in subnet1 `193.149.12.6 `. I do not see how I could fix this by changing routing entries. I would think when a host has an interface in a network, e.g. subnet1 `193.149.12.0/24`, the routing must always go through the address the host has in this network, `193.149.12.6` here.

So I would prefer a solution where the PVA gateway *always* sets the source IP address of the UDP reply packet to the IP address of the PVA gateway that received the name search request independently of any routing configuration.

I have tried to fix this myself in pvxs, but it is a bit challenging. Here is a link how to set the UDP source address in Linux:
  https://stackoverflow.com/questions/3062205/setting-the-source-ip-for-a-udp-socket

I have the impression that this change would be possible but it is not easy. The information of the *receiving* IP address on the PVA gateway is currently not available in function reply() in udp_collector.cpp where I would need it. I tried to add something like `SockAddr incoming_addr` to `struct PVXS_API Search` and set this in `UDPCollector::process_one() ` to `dest`. 
so it would be available in function `reply()`. I had to stop then, because it was getting too late...

**Comment by goetzpf** [comment]
I found a solution. The patch below solves my problem. Then the source IP of the UDP reply packet is correct.
A link to the patch file "pvxs-udp-src.patch" and the tshark capture file "raw-blc-2025-02-11.pcapng.gz" is here:

https://e.pcloud.link/publink/show?code=kZQM4nZXnBUxNjF1HzQvW9UD29fjpfi1Rvk

```diff
diff --git a/src/osiSockExt.h b/src/osiSockExt.h
index 6bd9ea8..46b4d0d 100644
--- a/src/osiSockExt.h
+++ b/src/osiSockExt.h
@@ -46,9 +46,8 @@ struct PVXS_API SockAddr {
         sockaddr_in6 in6;
 #endif
     };
-private:
-    store_t  store;
 public:
+    store_t  store;
 
     explicit SockAddr(int af = AF_UNSPEC);
     explicit SockAddr(const char *address, unsigned short port=0);
diff --git a/src/server.cpp b/src/server.cpp
index 0a45dc9..29c0402 100644
--- a/src/server.cpp
+++ b/src/server.cpp
@@ -747,7 +747,7 @@ void Server::Pvt::onSearch(const UDPManager::Search& msg)
     if(!M.good() || !H.good()) {
         log_crit_printf(serverio, "Logic error in Search buffer fill\n%s", "");
     } else {
-        (void)msg.reply(searchReply.data(), pktlen);
+        (void)msg.reply(searchReply.data(), pktlen, &(msg.incoming_addr));
     }
 }
 
diff --git a/src/udp_collector.cpp b/src/udp_collector.cpp
index 8e0366e..2c8fedd 100644
--- a/src/udp_collector.cpp
+++ b/src/udp_collector.cpp
@@ -88,7 +88,7 @@ struct UDPCollector final : public UDPManager::Search,
 
     // Search interface
 public:
-    virtual bool reply(const void *msg, size_t msglen) const override;
+    virtual bool reply(const void *msg, size_t msglen, const SockAddr *incoming=NULL) const override;
 };
 
 
@@ -272,6 +272,7 @@ bool UDPCollector::handle_one()
 
 void UDPCollector::process_one(const SockAddr &dest, const uint8_t *buf, size_t nrx, origin_t origin)
 {
+    this->incoming_addr= dest; // remember incoming address
     FixedBuf M(true, const_cast<uint8_t*>(buf), nrx);
     Header head{};
     from_wire(M, head); // overwrites M.be
@@ -478,17 +479,58 @@ void UDPCollector::forwardM(const SockAddr& origin, const uint8_t *pbuf, size_t
 
     sock.mcast_prep_sendto(lo_mcast_addr);
     src = lo_mcast_addr.addr;
-    reply(&buf[0], cmd_origin_tag_size+plen);
+    reply(&buf[0], cmd_origin_tag_size+plen, NULL);
 }
 
-bool UDPCollector::reply(const void *msg, size_t msglen) const
+bool UDPCollector::reply(const void *msg, size_t msglen, const SockAddr *incoming) const
 {
     manager->loop.assertInLoop();
 
     log_hex_printf(logio, Level::Debug, msg, msglen, "Send %s -> %s\n",
                    bind_addr.tostring().c_str(), src.tostring().c_str());
 
-    auto ntx = sendto(sock.sock, (char*)msg, msglen, 0, &src->sa, src.size());
+    ssize_t ntx;
+    if ((incoming==NULL) || (src.family()!=AF_INET))
+      {
+        ntx = sendto(sock.sock, (char*)msg, msglen, 0, &src->sa, src.size());
+      }
+    else
+      {
+        struct msghdr msgh;
+        //char msg_control[1024];
+        struct cmsghdr *cmsg;
+        //struct in_pktinfo *pktinfo;
+        struct in_pktinfo pktinfo;
+        struct iovec iov;
+
+        // Prepare IP_PKTINFO control message
+        memset(&pktinfo, 0, sizeof(pktinfo));
+        pktinfo.ipi_ifindex = 0;
+        pktinfo.ipi_spec_dst =  incoming->store.in.sin_addr;
+
+        // Prepare message header
+        memset(&msgh, 0, sizeof(msgh));
+        iov.iov_base = (void *)msg;
+        iov.iov_len = msglen;
+        msgh.msg_name= (void*)(&src->sa);
+        msgh.msg_namelen= src.size();
+        msgh.msg_iov = &iov;
+        msgh.msg_iovlen = 1;
+
+        // Allocate control message buffer
+        char control[CMSG_SPACE(sizeof(pktinfo))];
+        msgh.msg_control = control;
+        msgh.msg_controllen = sizeof(control);
+
+        // Add the IP_PKTINFO control message to the control buffer
+        cmsg = CMSG_FIRSTHDR(&msgh);
+        cmsg->cmsg_level = IPPROTO_IP;
+        cmsg->cmsg_type = IP_PKTINFO;
+        cmsg->cmsg_len = CMSG_LEN(sizeof(pktinfo));
+        memcpy(CMSG_DATA(cmsg), &pktinfo, sizeof(pktinfo));
+
+        ntx = sendmsg(sock.sock, &msgh, 0); // flags==0
+      }
     if(ntx<0) {
         int err = evutil_socket_geterror(sock.sock);
         if(err==SOCK_EWOULDBLOCK || err==EAGAIN || err==SOCK_EINTR) {
diff --git a/src/udp_collector.h b/src/udp_collector.h
index 9087934..b57cbf0 100644
--- a/src/udp_collector.h
+++ b/src/udp_collector.h
@@ -51,6 +51,7 @@ struct PVXS_API UDPManager
         std::vector<std::string> otherproto; // any protocols other than "tcp"
         SockAddr src;
         SockAddr server;
+        SockAddr incoming_addr; // set in process_one()
         uint32_t searchID;
         uint8_t peerVersion;
         bool protoTCP = false; // included protocol "tcp"
@@ -65,7 +66,7 @@ struct PVXS_API UDPManager
         decltype (names)::const_iterator begin() const { return names.begin(); }
         decltype (names)::const_iterator end() const   { return names.end(); }
 
-        virtual bool reply(const void *msg, size_t msglen) const =0;
+        virtual bool reply(const void *msg, size_t msglen, const SockAddr *incoming) const =0;
         Search() = default;
         Search(const Search&) = delete;
         Search& operator=(const Search&) = delete;
```

This is not a clean solution, for example it lacks IP6 support and making a property "store" of SockAddr public is a bit simple. But it shows, that in principle this works, at least with Linux.

**Comment by mdavidsaver** [comment]
@goetzpf Your diagram helps me.  Having it notated with IP addresses would help more.  I am also not clear host many host computers are involved.  Are the two CA gateways two processes on one host, or one process each on two hosts?

I see that there is an IP router which allows some traffic __around__ the subnet 1 gateway host.  I suspect this is the source of much of my confusion.  How is this configured?

fyi. looking again through `routing-gateway.txt` I notice what appears to be a conflict for `172.30.134.0/24`.  Is this by design?

```
172.30.134.0/24 dev vlan-devx proto kernel scope link src 172.30.134.7 
172.30.134.0/24 dev bond0 proto kernel scope link src 172.30.134.57 
```

**Comment by mdavidsaver** [comment]
> I found a solution. The patch below solves my problem.

This is the basic shape of a solution, which I see as worth pursing.

At the same time.  I am suspecting that there is an opportunity for a PVA version of this "hydra" system design to be at least somewhat simpler.  eg. I don't think there is a need for pairs of gateways and "ca net".  A single `p4p.gw` instance can have many client and servers, with arbitrary interconnects, so the equivalent of the "ca net" can be internal to the gateway process.

So I think you could compact the entire "hydra" into a single `p4p.gw` instance on a single host, provided that host had interfaces on all necessary subnets.  Now forcing all of this traffic through a single process may well be a performance bottleneck, although I have some ideas for dealing with this.

**Comment by goetzpf** [comment]
> [@goetzpf](https://github.com/goetzpf) Your diagram helps me. Having it notated with IP addresses would help more. I am also not clear how many host computers are involved. Are the two CA gateways two processes on one host, or one process each on two hosts?
In my test, the IP addresses are:
subnet1: 193.149.12.0/24, OPI Console: 193.149.12.42
ca-net: 192.168.16.0/24, Gateway IP (host gwc3c) here: 192.168.16.7
subnet2: 172.30.130.0/24, IOC IP 172.30.130.22

The two CA Gateways shown in the picture are there to illustrate that there is a CA gateway for each subnet. The UDP broadcaster is a program that runs on the host that also runs the CA gateway for the subnet.

In my test, the CA gateway between ca-net and subnet1 exists, but is not relevant for the shown connection between the OPI console and the IOC. The OPI console connects to the gateway with interfaces in the ca-net and subnet2, the one at the lower part of the picture.

In our original concept we planned to have virtual machines for each CA gateway. Later we found that a single Linux host with a "trunk" network line that has all VLANs on it, could run *all* CA gateway processes.

In the last year, a second gateway host together with a big number of new subnets was introduced. So we now have two CA gateway hosts, each connects to a large set of subnets, but these sets do not overlap.

Host "gwc3c" with an interface (among others) in ca-net (192.168.16.0/24) and subnet2 (172.30.130.0/24) has no interface in subnet1 (193.149.12.0/24). The older CA gateway host "gwc2c" has an interface in ca-net (192.168.16.0/24 with IP 192.168.16.6) and subnet1 (193.149.12.0/24 with IP 193.149.12.6). The UDP broadcaster on gwc2c forwards the UDP packet from the OPI console to the ca-net.
> 
> I see that there is an IP router which allows some traffic **around** the subnet 1 gateway host. I suspect this is the source of much of my confusion. How is this configured?
This just means that clients from *any* subnet can always connect to addresses in the ca-net. This is common IP routing, nothing special. 
> 
> fyi. looking again through `routing-gateway.txt` I notice what appears to be a conflict for `172.30.134.0/24`. Is this by design?
> 
> ```
> 172.30.134.0/24 dev vlan-devx proto kernel scope link src 172.30.134.7 
> 172.30.134.0/24 dev bond0 proto kernel scope link src 172.30.134.57 
> ```
I see this with "ip route":
```
cagwexp@gwc3c:~$ ip route | grep 172.30.134
default via 172.30.134.1 dev bond0 proto static 
172.30.134.0/24 dev bond0 proto kernel scope link src 172.30.134.57 
172.30.134.0/24 dev vlan-devx proto kernel scope link src 172.30.134.7 
```
Maybe this is a configuration error, but doesn't it simply mean that the host has two interfaces in the same network ? I do not administer the system but I could ask our sysadmin what this means.

**Comment by mdavidsaver** [comment]
> subnet1: 193.149.12.0/24, OPI Console: 193.149.12.42
> ca-net: 192.168.16.0/24, Gateway IP (host gwc3c) here: 192.168.16.7
> subnet2: 172.30.130.0/24, IOC IP 172.30.130.22

Ok.  I have been focusing on the wrong gateway host.  I have notated your diagram with what I think I understand.  What I don't understand is the location of `172.30.134.1`?  Does the IP router shown also have at interface with this IP?

Based on this diagram, I would expect that `gwc3c` would have its default route through `192.168.16.1`.  At the least, I think that adding a route to `193.149.12.0/24` via `192.168.16.1` would be one way to resolve your issues with p4p.gw (it may cause other problems...).

![Image](https://github.com/user-attachments/assets/cc40388f-17c3-4ea3-b105-64b1dd1c24d4)

**Comment by mdavidsaver** [comment]
```diff
+        // Prepare IP_PKTINFO control message
+        memset(&pktinfo, 0, sizeof(pktinfo));
+        pktinfo.ipi_ifindex = 0;
+        pktinfo.ipi_spec_dst =  incoming->store.in.sin_addr;
```

I think it will also be necessary to set `ipi_ifindex` to avoid the appearance of address spoofing (or actual spoofing by a malicious client).  This can be accomplished by `IfaceMap::index_of(const SockAddr)`, which will need to be added.  The necessary information is already stored, so this will only mean adding another "query" method.

**Comment by mdavidsaver** [comment]
@goetzpf Would you please look into testing https://github.com/epics-base/pvxs/pull/101 ?  To state the obvious, please be prepared for bugs...

**Comment by goetzpf** [comment]
Regarding your annotations to the diagram:
The interface of gwc3c in network 172.30.130.0/24 is 172.30.130.7, not 172.30.134.57.
Everything else is correct.

> What I don't understand is the location of 172.30.134.1? Does the IP router shown also have at interface with this IP?

This IP router is not shown in the picture as it is not needed for the channel access connection between the OPI console and the IOC.

> Based on this diagram, I would expect that gwc3c would have its default route through 192.168.16.1. At the least, I think that adding a route to 193.149.12.0/24 via 192.168.16.1 would be one way to resolve your issues with p4p.gw (it may cause other problems...).

You are correct that the default route on gwc3c should be 192.168.16.1. This is a misconfiguration. On our other gateway host, gwc2c, we have this fixed. Setting the default route to 192.168.16.1 would solve the problem only in the case the client is in a network where the gateway has no interface in.

But I ran the same test on gwc2c with IOC IP 192.168.21.142, client still with IP 193.149.12.42. The gateway has among others the IP addresses 193.149.12.6 and 192.168.16.6, the default route is via 192.168.16.1. In this case the gateway *has* an interface in the network of the client and it also doesn't work. The source IP of the UDP reply packet is 193.149.12.6 here although the gateway runs on 192.168.16.6. 

 I am not sure what would happen if we set a route to network 193.149.12.0/24 via 192.168.16.1, we haven't tested this. But how could the gateway find IOCs in network 193.149.12.0/24 with UDP broadcasts when the default route doesn't use the interface in that network ?

Regarding your last message: I will test this today.

**Comment by goetzpf** [comment]
> [@goetzpf](https://github.com/goetzpf) Would you please look into testing [epics-base/pvxs#101](https://github.com/epics-base/pvxs/pull/101) ? To state the obvious, please be prepared for bugs...

I tested your patched version of pvxs today and my simple test works with both our gateway hosts, thanks !
I am looking forward to have this integrated into the main repository.

**Comment by goetzpf** [comment]
Oh no !
Now the usual PV discovery is broken.
When I run on gwc2c (interface among others 192.168.16.6)
```
pvinfo idadm:calc1
```
I get:
```
Timeout
idadm:calc1 Error: channel destroyed
```
My original patch has the same problem.

**Comment by goetzpf** [comment]
My tests indicate this this error occurs, when you try to set a *broadcast* address as source address of the UDP packet.
It seems this is not allowed.
I patched my example so that in case the UDP source IP would be a broadcast address, it is replaced by the corresponding interface address of the network, e.g.:
`192.168.16.255:5075` is changed to `192.168.16.7:5075`, then everything works.
The client can read a PV with directed access by setting `EPICS_PVA_ADDR_LIST` and with access by UDP broadcast.

If think `IfaceMap` should be extended by a method that converts a broadcast address to the interface address, e.g. `192.168.16.255` becomes `192.168.16.7`, preferably with keeping the port number.

**Comment by goetzpf** [comment]
I think I found the problem in your version of pvxs. It is not the problem that I thought it was:
Here is the patch that makes it work:
```diff
diff -r 6b5f220b7e50 src/udp_collector.cpp
--- a/src/udp_collector.cpp	Wed Feb 19 22:24:07 2025 +0100
+++ b/src/udp_collector.cpp	Wed Feb 19 22:44:49 2025 +0100
@@ -332,13 +334,17 @@
 
         if(!M.good() || !(flags&pva_search_flags::Unicast) || dest.family()!=AF_INET) {
             // invalid, bcast, or not ipv4
-
+#if 0
             if(!(flags&pva_search_flags::Unicast) && origin!=OriginTag) {
                 // can't reply with non-ucast as source addr
                 dest = SockAddr(dest.family());
+                std::cerr << "UDPCollector::process_one CANT REPLY line " << __LINE__ << "\n"; //@@@
             }
 
         } else if(origin==Direct) {
+#else
+        } else if(dest.compare(lo_mcast_addr.addr,false)!=0) {
+#endif
             assert(buf==&this->buf[cmd_origin_tag_size]);
             // clear unicast flag in forwarded message
             *save_flags &= ~pva_search_flags::Unicast;
```
Somehow, when I make the PV name request the way it is usually done, with an UDP broadcast, the branch within `#if 0` is followed.
If I change the code here to be more like that from my older version of pvxs, it works for both cases. With directed access by setting `EPICS_PVA_ADDR_LIST` to ` 192.168.16.7` and with access by UDP broadcast.

**Comment by mdavidsaver** [comment]
Can you attach the routing table for `gwc2c`?

> Now the usual PV discovery is broken.

I am not able to reproduce.  If you don't see any error messages printed, please make a packet capture to see what is happening.

> Somehow, when I make the PV name request the way it is usually done, with an UDP broadcast, the branch within `#if 0` is followed.

A broadcast SEARCH message __must__ not have the Unicast flag set, so it should match the `!(flags&pva_search_flags::Unicast)` condition.

It should then also match the `origin!=OriginTag` condition since broadcasts should always be Direct (never forwarded).

This then clears the broadcast address stored in `dest` because, as you find, it is not permitted to use a broadcast address as a UDP source address.

This should leave `srcIface` set to a non-zero value.  So the reply `sendmsg()` should have a non-zero `ipi_ifindex`, but leave `ipi_addr` and `ipi_spec_dst` unset (implicitly `0.0.0.0`).

My understanding is that this causes the IP stack to select a source address based on a restricted routing table lookup based on the interface index.  This can be simulated with eg. `ip route get 1.1.1.1 oif br0`.

> convert a broadcast address to the interface address

I was hoping to avoid this lookup since it would need a special case to handle at least `255.255.255.255`.

**Comment by mdavidsaver** [comment]
> I was hoping to avoid this lookup since it would need a special case to handle at least `255.255.255.255`.

Another "special" case being multicast search.

fyi. multicast for search+beacons is my recommended alternative to your UDP re-broadcasting nodes.

**Comment by goetzpf** [comment]
>  Can you attach the routing table for gwc2c?

Here it is with the relevant interfaces, there are many more but none in the 172.30.130.0/24 network:
```
cagw@gwc2c:~$ ip route | grep '192.168.16\|193.149.12'
default via 192.168.16.1 dev bond1.467 onlink 
192.168.16.0/24 dev bond1.467 proto kernel scope link src 192.168.16.6 
193.149.12.0/24 dev bond1.400 proto kernel scope link src 193.149.12.6 
```

> I am not able to reproduce. If you don't see any error messages printed, please make a packet capture to see what is happening.

As I later found out, this was an error only printed by my test version of pvxs. In your version, this is not printed due to the `if` clause I changed later.

Of course, when I say "your version" I should clarify how I got the source, so we don't talk about different things:
```
gh repo clone epics-base/pvxs
cd pvxs
gh pr checkout 101
```

I saw now, that this version of pvxs *even doesn't work* when it is used as a port changing gateway.

When I create this docker network (strictly speaking I use podman, but this makes no difference here):
```
podman network create --subnet  10.90.0.0/24 --gateway 10.90.0.1 --ip-range 10.90.0.10-10.90.0.254 network0
```
Then run an IOC in docker with option ` --network network0 --ip 10.90.0.10` at address 10.90.0.10 and  start the IOC like this:
```
EPICS_PVAS_BROADCAST_PORT=6076  ./run-softioc.sh
```
so the Soft-IOC runs on port 6076.
Then I have this gateway configuration:
```
{
    "version":2,
    "clients":[
        {
            "name":"GWCLIENT1",
            "provider":"pva",
            "addrlist":"10.90.0.255",
            "autoaddrlist":false,
            "bcastport":6076
        }
    ],
    "servers":[
        {
            "name":"CASERVER1",
            "clients":["GWCLIENT1"],
            "interface":["10.90.0.100"],
            "addrlist":"",
            "#ignoreaddr":[""],
            "autoaddrlist":false,
            "serverport":5075,
            "bcastport":5076,
            "getholdoff":0,
            "statusprefix":"PV:",
            "access":"",
            "pvlist":""
        }
    ]
}
```
I run the PVA gateway in docker with IP address 10.90.0.100 and start it.

Now I start a third docker container on address 10.90.0.16 and enter:
```
pvinfo goetz:calc1
```
then I get:
```
Timeout
goetz:calc1 Error: channel destroyed
```

I must admit, that the patch I posted here yesterday was incomplete, this is the complete patch that fixes the problem:

```diff
diff --git a/src/evhelper.cpp b/src/evhelper.cpp
index ab98f99..1ba3d25 100644
--- a/src/evhelper.cpp
+++ b/src/evhelper.cpp
@@ -885,6 +885,14 @@ bool IfaceMap::is_broadcast(const SockAddr& addr)
     });
 }
 
+SockAddr IfaceMap::broadcast_to_ifaddr(const SockAddr& addr)
+{
+    if (!is_broadcast(addr))
+      return addr;
+    std::string ifname= name_of(addr);
+    return address_of(ifname);
+}
+
 SockAddr IfaceMap::address_of(const std::string& name)
 {
     Guard G(lock);
diff --git a/src/evhelper.h b/src/evhelper.h
index 6a9904a..d2cf6ba 100644
--- a/src/evhelper.h
+++ b/src/evhelper.h
@@ -324,6 +324,10 @@ struct PVXS_API IfaceMap {
     bool is_lo(uint64_t index);
     // is this a valid interface or broadcast address?
     bool is_broadcast(const SockAddr& addr);
+
+    // try to get interface addr from broadcast addr:
+    SockAddr broadcast_to_ifaddr(const SockAddr& addr);
+
     // look up interface address.  useful for IPV4.  returns AF_UNSPEC if not found
     SockAddr address_of(const std::string& name);
     // all interface names except LO
diff --git a/src/server.cpp b/src/server.cpp
index 0a45dc9..2e9ab4b 100644
--- a/src/server.cpp
+++ b/src/server.cpp
@@ -4,7 +4,6 @@
  * in file LICENSE that is included with this distribution.
  */
 
-
 #include <list>
 #include <map>
 #include <system_error>
diff --git a/src/udp_collector.cpp b/src/udp_collector.cpp
index bfb2bc1..999184d 100644
--- a/src/udp_collector.cpp
+++ b/src/udp_collector.cpp
@@ -328,13 +328,7 @@ void UDPCollector::process_one(const uint8_t *buf, size_t nrx, origin_t origin)
 
         if(!M.good() || !(flags&pva_search_flags::Unicast) || dest.family()!=AF_INET) {
             // invalid, bcast, or not ipv4
-
-            if(!(flags&pva_search_flags::Unicast) && origin!=OriginTag) {
-                // can't reply with non-ucast as source addr
-                dest = SockAddr(dest.family());
-            }
-
-        } else if(origin==Direct) {
+        } else if(dest.compare(lo_mcast_addr.addr,false)!=0) {
             assert(buf==&this->buf[cmd_origin_tag_size]);
             // clear unicast flag in forwarded message
             *save_flags &= ~pva_search_flags::Unicast;
@@ -516,7 +510,14 @@ bool UDPCollector::reply(const void *msg, size_t msglen) const
                    bind_addr.tostring().c_str(), src.tostring().c_str(),
                    dest.tostring().c_str(), unsigned(srcIface));
 
-    auto ntx = sendtox{sock.sock, (char*)msg, msglen, &src, &dest, srcIface}.call();
+    SockAddr const *mydest= &dest;
+    SockAddr if_addr= manager->ifmap.broadcast_to_ifaddr(dest);
+    if (if_addr != dest) {
+        if_addr.setPort(dest.port());
+        mydest= &if_addr;
+    }
+
+    auto ntx = sendtox{sock.sock, (char*)msg, msglen, &src, mydest, srcIface}.call();
     if(ntx<0) {
         int err = evutil_socket_geterror(sock.sock);
         if(err==SOCK_EWOULDBLOCK || err==EAGAIN || err==SOCK_EINTR) {
```

When I run exactly the same test but this time with a version of pvxs where the patch above is appied, it works. I get this message:
```[root@ GPTESTS]$./run-pvinfo.sh
goetz:calc1
Server: 10.90.0.100:5075
Type:
    epics:nt/NTScalar:1.0
        double value
        alarm_t alarm
            int severity
            int status
            string message
        structure timeStamp
            long secondsPastEpoch
            int nanoseconds
            int userTag
        structure display
            double limitLow
            double limitHigh
            string description
            string units
            int precision
            enum_t form
                int index
                string[] choices
        control_t control
            double limitLow
            double limitHigh
            double minStep
        valueAlarm_t valueAlarm
            boolean active
            double lowAlarmLimit
            double lowWarningLimit
            double highWarningLimit
            double highAlarmLimit
            int lowAlarmSeverity
            int lowWarningSeverity
            int highWarningSeverity
            int highAlarmSeverity
            byte hysteresis

```

So it is reproducible on a single PC (currently I run this on my Linux laptop) that UDP broadcasts with pvxs `Sendto pktinfo#101` get no answer,  without anything funny going on regarding special routing entries or UDP-broadcaster processes. 

I saw the same yesterday with a "real world" example running on eds02, gwc3c and gwc2c as a client, only standard PVA ports in this case.  I used gwc2 as a client with `EPICS_PVA_AUTO_ADDR_LIST=NO EPICS_PVA_ADDR_LIST=192.168.16.255 pvinfo idadm:calc` so our UDP-broadcaster and the other network interfaces of gwc2c wouldn't be involved. Since the PVA server IP is on gwc3c has IP 192.168.16.7 this should have worked, too.

So
```
if(!(flags&pva_search_flags::Unicast) && origin!=OriginTag)
```
is somehow true causing `reply` never to be called. I saw that this line was added recently.

The part that changes the receiving address from a broadcast to an interface address  `broadcast_to_ifaddr()` is also necessary. If this is removed, your version of the PVA gateway, when only the `if` statement is changed, also prints `UDP TX Error on UDP 0.0.0.0:5076 -> .... : (101) Network is unreachable`.

**Comment by goetzpf** [comment]
Hello @mdavidsaver , 
have you had time to look into this issue with p4p (actually pvxs) ?
The latest patch here seems to work for me, but it would be nice to have
an official fix.

**Comment by mdavidsaver** [comment]
@goetzpf I updated https://github.com/epics-base/pvxs/pull/101 on friday if you have time to retest.

**Comment by goetzpf** [comment]
@mdavidsaver : I have tested your latest changes but the pva gateway doesn't reply to common UDP broadcasts. However, after I apply this change:
```diff
diff --git a/src/udp_collector.cpp b/src/udp_collector.cpp
index ad67f1c..716675c 100644
--- a/src/udp_collector.cpp
+++ b/src/udp_collector.cpp
@@ -289,6 +289,7 @@ bool UDPCollector::handle_one(const IfaceMap::Current& ifinfo)
         if(ifit!=srcIface->bcast.end()) {
             // dest is bcast, so replace with associated iface address
             dest = ifit->second;
+            origin = Forwarding;
 
         } else if((ifit=srcIface->addrs.find(dest))!=srcIface->addrs.end()) {
             // dest is interface address.  Nothing to do.
@@ -360,7 +361,7 @@ void UDPCollector::process_one(const uint8_t *buf, size_t nrx, origin_t origin,
         if(!M.good())
             return;
 
-        if(!(flags&pva_search_flags::Unicast) || dest.family()!=AF_INET) {
+        if((!(flags&pva_search_flags::Unicast) || dest.family()!=AF_INET) && (origin!=Forwarding)) {
             // invalid, bcast, or not ipv4
 
             if(!(flags&pva_search_flags::Unicast) && origin!=OriginTag) {
```
everything works for me, all my tests pass (UDP broadcast and unicast).
You have to ensure that the `if` block at line 360 is not executed when a UDP broadcast is received.

**Comment by mdavidsaver** [comment]
```diff
             // dest is bcast, so replace with associated iface address
             dest = ifit->second;
+            origin = Forwarding;
```

Unfortunately, broadcasts should not be forwarded as they are already received by all peers.

> You have to ensure that the `if` block at line 360 is not executed when a UDP broadcast is received.

This test is intended to prevent attempts to send a packet with a broadcast address as the __source__ address.  To my knowledge this is not permitted.

**Comment by goetzpf** [comment]
@mdavidsaver :  I have tested your most recent changes but the pva gateway still doesn't reply to UDP broadcasts. The following change:
```diff
diff --git a/src/udp_collector.cpp b/src/udp_collector.cpp
index ef9e354..276b382 100644
--- a/src/udp_collector.cpp
+++ b/src/udp_collector.cpp
@@ -290,6 +290,7 @@ bool UDPCollector::handle_one(const IfaceMap::Current& ifinfo)
         if(ifit!=srcIface->bcast.end()) {
             // dest is bcast, so replace with associated iface address
             dest = ifit->second;
+            origin = Forwarding;
 
         } else if((ifit=srcIface->addrs.find(dest))!=srcIface->addrs.end()) {
             // dest is interface address.  Nothing to do.
```
is needed to make it work. This version passes all my tests.

**Comment by mdavidsaver** [comment]
I don't understand what is going on.  @goetzpf could you elaborate or provide wireshark captures for both modified and unmodified?

**Comment by goetzpf** [comment]
In file src/udp_collector.cpp:

In line 275 origin is set to Broadcast:
```
    origin_t origin = Broadcast;
```

In line 292, dest ist changed:
```
            dest = ifit->second;
```
but origin has still the value `Broadcast`.

In line 320 `process_one` is called, parameter `origin` is still `Broadcast`.
```
    process_one(rxbuf, nrx, origin, ifinfo);
```

In this function the condition in line 377:
```
        if(origin==Broadcast || dest.family()!=AF_INET) {
            // invalid, bcast, or not ipv4
```
is true, since `origin` is still `Broadcast`. As your comment suggests, this
request is then ignored. 

This is wrong since this happens when a client does a name resulution request with a udp broadcast.

WHen you append this to line 292:
```
            origin = Forwarding;
```
this fixes the problem since the condition in line 377 is no longer true. 

Of course, you could also change line 377 and 378 to:
```
        if(dest.family()!=AF_INET) {
            // invalid or not ipv4
```

I have caputured the traffic in my experiment with wireshark with the version that doesn't work, here is the link:

https://e.pcloud.link/publink/show?code=XZH5zqZIryhY1bG4mmU8ax6lqLhTjT2nXHX

For this experiment, the gateway has addess 192.168.16.6, the client has address 192.168.16.5. You see that the UDP broadcast doesn't get a reply.

**Comment by mdavidsaver** [comment]
> In line 292, dest ist changed:
> 
> ```
>             dest = ifit->second;
> ```

From your capture, I expect that this should be replacing `192.168.16.255` with `192.168.16.7`, correct?



> but origin has still the value `Broadcast`.

Correct, as this packet originated as a broadcast.


> is true, since `origin` is still `Broadcast`. As your comment suggests, this
> request is then ignored.

I do not think this is a correct interpretation.  The logic here is either that the packet will be forwarded (via. local mcast) or processed locally.  In no case is the message simply ignored.

So I can think of two possibilities.  Either the gateway code is choosing to ignore this search (eg. incorrect values pass to ACF hostname rules?), or the reply `sendmsg()` is made with incorrect, but valid, arguments.  (you did not mention seeing any "UDP TX Error" log messages, so I assume validity)

The first test I would ask is to enable logging (`export PVXS_LOG=pvxs.udp.io=DEBUG`) and report if you see the "Send" message on entry to `UDPCollector::reply()`.

Another suggestion would be to run run an IOC on this gateway (gwc3c) using the same PVA configuration as the gateway.

**Comment by goetzpf** [comment]
I now have a very simple setup to reproduce the error. If you want to try this yourself, you have only to change the IP addresses in the gateway configuration file.

# Test setup

## IOC

I run a simple soft-ioc that was created with:

```
makeBaseApp.pl -b "$EPICS_BASE" -t example "$IOC_NAME"
makeBaseApp.pl -i -t example -p "$IOC_NAME" "ioc$IOC_NAME"
```

with this command:
```
EPICS_CA_SERVER_PORT=6064 EPICS_PVAS_BROADCAST_PORT=6076 ./st.cmd
```

## PVA-Gateway

I run a pva gateway with this configuration:

```
{
    "version":2,
    "clients":[
        {
            "name":"GWCLIENT1",
            "provider":"pva",
            "addrlist":"192.168.178.255",
            "autoaddrlist":false,
            "bcastport":6076
        }
    ],
    "servers":[
        {
            "name":"CASERVER1",
            "clients":["GWCLIENT1"],
            "interface":["192.168.178.94"],
            "addrlist":"",
            "#ignoreaddr":[""],
            "autoaddrlist":false,
            "serverport":5075,
            "bcastport":5076,
            "getholdoff":0,
            "statusprefix":"PV:",
            "access":"",
            "pvlist":""
        }
    ]
}
```

## Client

I run pvinfo:

```
pvinfo $(whoami):calc1I
```

# Test results

Without this patch:

```diff
diff -r a638bd87ea1c -r 2207491bd469 src/udp_collector.cpp
--- a/src/udp_collector.cpp     Wed Apr 23 16:03:43 2025 +0200
+++ b/src/udp_collector.cpp     Wed May 07 15:25:52 2025 +0200
@@ -290,6 +290,7 @@
         if(ifit!=srcIface->bcast.end()) {
             // dest is bcast, so replace with associated iface address
             dest = ifit->second;
+            origin = Forwarding;
 
         } else if((ifit=srcIface->addrs.find(dest))!=srcIface->addrs.end()) {
             // dest is interface address.  Nothing to do.
```

it doesn't work.

# Debug output from the PVA gateway

When, as you suggested, I set this environment variable before starting the gateway::

```
PVXS_LOG=pvxs.udp.io=DEBUG
```

I get more debug messages.

Here are the messages for the case that doesn't work:

```
INFO:p4p.gw:*** Gateway STARTS now using "testgw-general-port.json".
INFO:p4p.gw:Client effective configuration for GWCLIENT1:
INFO:p4p.gw:    EPICS_PVA_ADDR_LIST : 192.168.178.255
INFO:p4p.gw:    EPICS_PVA_AUTO_ADDR_LIST : NO
INFO:p4p.gw:    EPICS_PVA_BROADCAST_PORT : 6076
2025-05-07T17:10:25.200359039 WARN pvxs.tcp.setup Server unable to bind port 5075, falling back to 192.168.178.94:43865
INFO:p4p.gw:Server effective configuration for CASERVER1_0:
INFO:p4p.gw:    EPICS_PVAS_AUTO_BEACON_ADDR_LIST : NO
INFO:p4p.gw:    EPICS_PVAS_BEACON_ADDR_LIST : 
INFO:p4p.gw:    EPICS_PVAS_BROADCAST_PORT : 5076
INFO:p4p.gw:    EPICS_PVAS_IGNORE_ADDR_LIST : 
INFO:p4p.gw:    EPICS_PVAS_INTF_ADDR_LIST : 192.168.178.94:5075
INFO:p4p.gw:    EPICS_PVAS_SERVER_PORT : 43865
INFO:p4p.gw:    EPICS_PVA_ADDR_LIST : 192.168.178.94:5075
INFO:p4p.gw:    EPICS_PVA_AUTO_ADDR_LIST : NO
INFO:p4p.gw:    EPICS_PVA_BROADCAST_PORT : 5076
INFO:p4p.gw:    EPICS_PVA_CONN_TMO : 30
INFO:p4p.gw:    EPICS_PVA_INTF_ADDR_LIST : 192.168.178.94:5075
INFO:p4p.gw:    EPICS_PVA_NAME_SERVERS : 
INFO:p4p.gw:    EPICS_PVA_SERVER_PORT : 43865
INFO:p4p.gw:Status PV: PV:StatsTime
INFO:p4p.gw:Status PV: PV:asDebug
INFO:p4p.gw:Status PV: PV:asTest
INFO:p4p.gw:Status PV: PV:cache
INFO:p4p.gw:Status PV: PV:clients
INFO:p4p.gw:Status PV: PV:ds:byhost:rx
INFO:p4p.gw:Status PV: PV:ds:byhost:tx
INFO:p4p.gw:Status PV: PV:ds:bypv:rx
INFO:p4p.gw:Status PV: PV:ds:bypv:tx
INFO:p4p.gw:Status PV: PV:poke
INFO:p4p.gw:Status PV: PV:refs
INFO:p4p.gw:Status PV: PV:stats
INFO:p4p.gw:Status PV: PV:threads
INFO:p4p.gw:Status PV: PV:us:byhost:rx
INFO:p4p.gw:Status PV: PV:us:byhost:tx
INFO:p4p.gw:Status PV: PV:us:bypv:rx
INFO:p4p.gw:Status PV: PV:us:bypv:tx
INFO:p4p.gw:Setup GW clients to ignore GW servers
2025-05-07T17:10:26.813450490 DEBUG pvxs.udp.io UDP 0x7f14300018a0 event 2
0000 : CA024000 27000000 D3771B68 00000000
0010 : 3AC97030 00010000 00000000 00000000
0020 : 0000FFFF 00000000 D3130374 6370FF
2025-05-07T17:10:26.813576756 DEBUG pvxs.udp.io UDP Rx 47, 192.168.178.94:44905 -> 192.168.178.94 @3 (0.0.0.0:6076) : orig 0
2025-05-07T17:10:27.384995401 DEBUG pvxs.udp.io UDP 0x7f1430002270 event 2
0000 : CA020003 31000000 01000000 00000000
0010 : 00000000 00000000 0000FFFF 00000000
0020 : AD920103 74637001 00413020 100B676F
0030 : 65747A3A 63616C63 31 
2025-05-07T17:10:27.385061525 DEBUG pvxs.udp.io UDP Rx 57, 192.168.178.94:37549 -> 192.168.178.94 @3 (0.0.0.0:5076) : orig 0
2025-05-07T17:10:27.629953757 DEBUG pvxs.udp.io UDP 0x7f1430002270 event 2
0000 : CA020003 31000000 02000000 00000000
0010 : 00000000 00000000 0000FFFF 00000000
0020 : AD920103 74637001 00413020 100B676F
0030 : 65747A3A 63616C63 31 
2025-05-07T17:10:27.630010664 DEBUG pvxs.udp.io UDP Rx 57, 192.168.178.94:37549 -> 192.168.178.94 @3 (0.0.0.0:5076) : orig 0
2025-05-07T17:10:28.119286078 DEBUG pvxs.udp.io UDP 0x7f1430002270 event 2
0000 : CA020003 31000000 03000000 00000000
0010 : 00000000 00000000 0000FFFF 00000000
0020 : AD920103 74637001 00413020 100B676F
0030 : 65747A3A 63616C63 31 
2025-05-07T17:10:28.119323468 DEBUG pvxs.udp.io UDP Rx 57, 192.168.178.94:37549 -> 192.168.178.94 @3 (0.0.0.0:5076) : orig 0
2025-05-07T17:10:29.098195858 DEBUG pvxs.udp.io UDP 0x7f1430002270 event 2
0000 : CA020003 31000000 04000000 00000000
0010 : 00000000 00000000 0000FFFF 00000000
0020 : AD920103 74637001 00413020 100B676F
0030 : 65747A3A 63616C63 31 
2025-05-07T17:10:29.098251241 DEBUG pvxs.udp.io UDP Rx 57, 192.168.178.94:37549 -> 192.168.178.94 @3 (0.0.0.0:5076) : orig 0
2025-05-07T17:10:31.056305916 DEBUG pvxs.udp.io UDP 0x7f1430002270 event 2
0000 : CA020003 31000000 05000000 00000000
0010 : 00000000 00000000 0000FFFF 00000000
0020 : AD920103 74637001 00413020 100B676F
0030 : 65747A3A 63616C63 31 
2025-05-07T17:10:31.056355228 DEBUG pvxs.udp.io UDP Rx 57, 192.168.178.94:37549 -> 192.168.178.94 @3 (0.0.0.0:5076) : orig 0
```

And here are the messages for the case that does work, with the patch:

```
INFO:p4p.gw:*** Gateway STARTS now using "testgw-general-port.json".
INFO:p4p.gw:Client effective configuration for GWCLIENT1:
INFO:p4p.gw:    EPICS_PVA_ADDR_LIST : 192.168.178.255
INFO:p4p.gw:    EPICS_PVA_AUTO_ADDR_LIST : NO
INFO:p4p.gw:    EPICS_PVA_BROADCAST_PORT : 6076
2025-05-07T17:11:57.182997020 WARN pvxs.tcp.setup Server unable to bind port 5075, falling back to 192.168.178.94:34083
INFO:p4p.gw:Server effective configuration for CASERVER1_0:
INFO:p4p.gw:    EPICS_PVAS_AUTO_BEACON_ADDR_LIST : NO
INFO:p4p.gw:    EPICS_PVAS_BEACON_ADDR_LIST : 
INFO:p4p.gw:    EPICS_PVAS_BROADCAST_PORT : 5076
INFO:p4p.gw:    EPICS_PVAS_IGNORE_ADDR_LIST : 
INFO:p4p.gw:    EPICS_PVAS_INTF_ADDR_LIST : 192.168.178.94:5075
INFO:p4p.gw:    EPICS_PVAS_SERVER_PORT : 34083
INFO:p4p.gw:    EPICS_PVA_ADDR_LIST : 192.168.178.94:5075
INFO:p4p.gw:    EPICS_PVA_AUTO_ADDR_LIST : NO
INFO:p4p.gw:    EPICS_PVA_BROADCAST_PORT : 5076
INFO:p4p.gw:    EPICS_PVA_CONN_TMO : 30
INFO:p4p.gw:    EPICS_PVA_INTF_ADDR_LIST : 192.168.178.94:5075
INFO:p4p.gw:    EPICS_PVA_NAME_SERVERS : 
INFO:p4p.gw:    EPICS_PVA_SERVER_PORT : 34083
INFO:p4p.gw:Status PV: PV:StatsTime
INFO:p4p.gw:Status PV: PV:asDebug
INFO:p4p.gw:Status PV: PV:asTest
INFO:p4p.gw:Status PV: PV:cache
INFO:p4p.gw:Status PV: PV:clients
INFO:p4p.gw:Status PV: PV:ds:byhost:rx
INFO:p4p.gw:Status PV: PV:ds:byhost:tx
INFO:p4p.gw:Status PV: PV:ds:bypv:rx
INFO:p4p.gw:Status PV: PV:ds:bypv:tx
INFO:p4p.gw:Status PV: PV:poke
INFO:p4p.gw:Status PV: PV:refs
INFO:p4p.gw:Status PV: PV:stats
INFO:p4p.gw:Status PV: PV:threads
INFO:p4p.gw:Status PV: PV:us:byhost:rx
INFO:p4p.gw:Status PV: PV:us:byhost:tx
INFO:p4p.gw:Status PV: PV:us:bypv:rx
INFO:p4p.gw:Status PV: PV:us:bypv:tx
INFO:p4p.gw:Setup GW clients to ignore GW servers
2025-05-07T17:11:58.601892376 DEBUG pvxs.udp.io UDP 0x7f0d6c002270 event 2
0000 : CA020003 31000000 01000000 00000000
0010 : 00000000 00000000 0000FFFF 00000000
0020 : 77CB0103 74637001 00413020 100B676F
0030 : 65747A3A 63616C63 31 
2025-05-07T17:11:58.602008643 DEBUG pvxs.udp.io UDP Rx 57, 192.168.178.94:52087 -> 192.168.178.94 @3 (0.0.0.0:5076) : orig 1
2025-05-07T17:11:58.602036856 DEBUG pvxs.udp.io Forward as originated for 192.168.178.94
0000 : CA028016 00000010 00000000 00000000
0010 : 0000FFFF C0A8B25E CA020003 31000000
0020 : 01000000 00000000 00000000 00000000
0030 : 0000FFFF C0A8B25E 77CB0103 74637001
...
2025-05-07T17:11:58.602045803 DEBUG pvxs.udp.io Send 0.0.0.0:5076 -> 224.0.0.128:5076, 0.0.0.0,N/A
0000 : CA028016 00000010 00000000 00000000
0010 : 0000FFFF C0A8B25E CA020003 31000000
0020 : 01000000 00000000 00000000 00000000
0030 : 0000FFFF C0A8B25E 77CB0103 74637001
...
2025-05-07T17:11:58.602099282 DEBUG pvxs.udp.io UDP Rx 81, <> -> <> @1 (0.0.0.0:5076) : orig 2
2025-05-07T17:11:58.602111656 DEBUG pvxs.udp.io Ignore as originated for 192.168.178.94:5076
2025-05-07T17:11:58.612576097 DEBUG pvxs.udp.io UDP 0x7f0d6c0018a0 event 2
0000 : CA028003 00000031 66696E64 00000000
0010 : 00000000 00000000 00000000 00000000
0020 : AD760103 74637000 01123456 780B676F
0030 : 65747A3A 63616C63 31 
2025-05-07T17:11:58.612660345 DEBUG pvxs.udp.io UDP Rx 57, 192.168.178.94:44406 -> 192.168.178.94 @3 (0.0.0.0:6076) : orig 1
2025-05-07T17:11:58.612677968 DEBUG pvxs.udp.io Forward as originated for 192.168.178.94
0000 : CA028016 00000010 00000000 00000000
0010 : 0000FFFF C0A8B25E CA028003 00000031
0020 : 66696E64 00000000 00000000 00000000
0030 : 0000FFFF C0A8B25E AD760103 74637000
...
2025-05-07T17:11:58.612685281 DEBUG pvxs.udp.io Send 0.0.0.0:6076 -> 224.0.0.128:6076, 0.0.0.0,N/A
0000 : CA028016 00000010 00000000 00000000
0010 : 0000FFFF C0A8B25E CA028003 00000031
0020 : 66696E64 00000000 00000000 00000000
0030 : 0000FFFF C0A8B25E AD760103 74637000
...
2025-05-07T17:11:58.612735145 DEBUG pvxs.udp.io UDP Rx 81, <> -> <> @1 (0.0.0.0:6076) : orig 2
2025-05-07T17:11:58.612747437 DEBUG pvxs.udp.io Ignore as originated for 192.168.178.94:6076
2025-05-07T17:11:58.849655867 DEBUG pvxs.udp.io UDP 0x7f0d6c002270 event 2
0000 : CA020003 31000000 02000000 00000000
0010 : 00000000 00000000 0000FFFF 00000000
0020 : 77CB0103 74637001 00413020 100B676F
0030 : 65747A3A 63616C63 31 
2025-05-07T17:11:58.849734614 DEBUG pvxs.udp.io UDP Rx 57, 192.168.178.94:52087 -> 192.168.178.94 @3 (0.0.0.0:5076) : orig 1
2025-05-07T17:11:58.849776893 DEBUG pvxs.udp.io Forward as originated for 192.168.178.94
0000 : CA028016 00000010 00000000 00000000
0010 : 0000FFFF C0A8B25E CA020003 31000000
0020 : 02000000 00000000 00000000 00000000
0030 : 0000FFFF C0A8B25E 77CB0103 74637001
...
2025-05-07T17:11:58.849789096 DEBUG pvxs.udp.io Send 0.0.0.0:5076 -> 224.0.0.128:5076, 0.0.0.0,N/A
0000 : CA028016 00000010 00000000 00000000
0010 : 0000FFFF C0A8B25E CA020003 31000000
0020 : 02000000 00000000 00000000 00000000
0030 : 0000FFFF C0A8B25E 77CB0103 74637001
...
2025-05-07T17:11:58.849843858 DEBUG pvxs.udp.io UDP Rx 81, <> -> <> @1 (0.0.0.0:5076) : orig 2
2025-05-07T17:11:58.849853055 DEBUG pvxs.udp.io Ignore as originated for 192.168.178.94:5076
0000 : CA02C004 0000002D FE665313 3BB260C2
0010 : 10762B18 00000002 00000000 00000000
0020 : 0000FFFF 00000000 85230374 63700100
0030 : 01102030 41  
2025-05-07T17:11:58.850032661 DEBUG pvxs.udp.io Send 0.0.0.0:5076 -> 192.168.178.94:52087, 192.168.178.94:5076,wlp4s0

```
One difference I see in these two versions is that the second one, for the working pva gateway version, at some place says "Send". This cannot be found in the non-working version.

**Comment by mdavidsaver** [comment]
> ... I had the impression you were not content with this.  ...

No, I am not.  I have tried to expand on this here: https://github.com/epics-base/pvxs/pull/127#issuecomment-3067373587
